rustls/src/client/client_conn.rs

```diff
@@ -115,6 +115,17 @@ pub trait ResolvesClientCert: Send + Sync {
     fn has_certs(&self) -> bool;
 }
 
+/// A trait providing the ability to camouflage ClientHello to resist fingerprinting.
+/// Notice: wrongly configured cipher suites and extensions may result in failure in connection.
+pub trait ClientHelloCamouflager: Send + Sync {
+    /// Transform cipher suites created from rustls to camouflaged cipher suites.
+    fn transform_cipher_suites(&self, cipher_suites: Vec<CipherSuite>) -> Vec<CipherSuite>;
+
+    /// Transform extensions created from rustls to camouflaged extensions. Users may need to
+    /// reuse some provided extensions for a successful connection.
+    fn transform_extensions(&self, extensions: Vec<ClientExtension>) -> Vec<ClientExtension>;
```

Comment:
IIRC some extensions have to be a in a particular order to work. This interface would make it pretty easy to get that wrong.

Reply:
According to RFC5246#7.4.1.4, extensions MAY appear in any order in TLS 1.2 , while TLS 1.3 has an exception of pre_shared_key which must be the last if exists. I am not convinced that we should perform this check, because I believe only users who know what they are doing should use this feature, and they have more ways to mess up anything if they want.